File: typescript/rna-transcription/rna-transcription.ts
type NucleotideComplement = "C" | "G" | "A" | "U"

interface NucleotideComplementDictionary {
  readonly [key: string]: NucleotideComplement
}

type NucleotidesToRnaReducer = (rna: string, nucleotide: string) => string

const nucleotideComplements: NucleotideComplementDictionary = {
  G: "C",
  C: "G",
  T: "A",
  A: "U",
}

export function toRna(dnaStrand: string): string {
  const reducer: NucleotidesToRnaReducer = (rna, nucleotide) => {
    const complement = nucleotideComplements[nucleotide]
    if (!complement) {
      throw new TypeError(`Invalid input DNA.`)
    }

    return `${rna}${complement}`
  }

  return dnaStrand.split("").reduce(reducer, "")
}

File: clojure/rna-transcription/src/rna_transcription.clj
(ns rna-transcription)

(def nucleotide-complements
  (hash-map \G \C
            \C \G
            \T \A
            \A \U))

(defn get-nucleotide-complement
  "Return complement for nucleotide."
  [nucleotide]
  {:post [(some? %)]}
  (nucleotide-complements nucleotide))

(defn to-rna
  "Return RNA complement for DNA strand."
  [dna-strand]
  (apply str (map get-nucleotide-complement dna-strand)))